misc/rb.c

```diff
@@ -0,0 +1,495 @@
+ * stack, modifying the stack accordingly. Does not change the color
+ * of either node.  That is, it effects the following transition (or
+ * its mirror if N is on the other side of P, of course):
+ *
+ *    P          N
+ *  N  c  -->  a   P
+ * a b            b c
+ *
+ */
+static void rotate(struct rbnode **stack, int stacksz)
+{
+	CHECK(stacksz >= 2);
+
+	struct rbnode *parent = stack[stacksz - 2];
+	struct rbnode *child = stack[stacksz - 1];
+	int side = get_side(parent, child);
+	struct rbnode *a = get_child(child, side);
+	struct rbnode *b = get_child(child, !side);
+
+	if (stacksz >= 3) {
+		struct rbnode *grandparent = stack[stacksz - 3];
+
+		set_child(grandparent, get_side(grandparent, parent), child);
+	}
+
+	set_child(child, side, a);
+	set_child(child, !side, parent);
+	set_child(parent, side, b);
+	stack[stacksz - 2] = child;
+	stack[stacksz - 1] = parent;
+}
+
+/* The node at the top of the provided stack is red, and its parent is
+ * too.  Iteratively fix the tree so it becomes a valid red black tree
+ * again
+ */
+static void fix_extra_red(struct rbnode **stack, int stacksz)
+{
+	while (stacksz > 1) {
+		struct rbnode *node = stack[stacksz - 1];
+		struct rbnode *parent = stack[stacksz - 2];
+
+		/* Correct child colors are a precondition of the loop */
+		CHECK(!get_child(node, 0) || is_black(get_child(node, 0)));
+		CHECK(!get_child(node, 1) || is_black(get_child(node, 1)));
+
+		if (is_black(parent)) {
+			return;
+		}
+
+		/* We are guaranteed to have a grandparent if our
+		 * parent is red, as red nodes cannot be the root
+		 */
+		CHECK(stacksz >= 2);
+
+		struct rbnode *grandparent = stack[stacksz - 3];
+		int side = get_side(grandparent, parent);
+		struct rbnode *aunt = get_child(grandparent, !side);
+
+		if (aunt && is_red(aunt)) {
+			set_color(grandparent, RED);
+			set_color(parent, BLACK);
+			set_color(aunt, BLACK);
+
+			/* We colored the grandparent red, which might
+			 * have a red parent, so continue iterating
+			 * from there */
+			stacksz -= 2;
+			continue;
+		}
+
+		/* We can rotate locally to fix the whole tree.  First
+		 * make sure that node is on the same side of parent
+		 * as parent is of grandparent.
+		 */
+		int parent_side = get_side(parent, node);
+
+		if (parent_side != side) {
+			rotate(stack, stacksz);
+			node = stack[stacksz - 1];
+		}
+
+		/* Rotate the grandparent with parent, swapping colors */
+		rotate(stack, stacksz - 1);
+		set_color(stack[stacksz - 3], BLACK);
+		set_color(stack[stacksz - 2], RED);
+		return;
+	}
+
+	/* If we exit the loop, it's because our node is now the root,
+	 * which must be black */
+	set_color(stack[0], BLACK);
+}
+
+void rb_insert(struct rbtree *tree, struct rbnode *node)
+{
+	set_child(node, 0, NULL);
+	set_child(node, 1, NULL);
+
+	if (!tree->root) {
+		tree->root = node;
+		tree->max_depth = 1;
+		set_color(node, BLACK);
+		return;
+	}
+
+	struct rbnode *stack[tree->max_depth + 1];
```

Comment:
I think checkpatch may barf on this, the coding style really wants all declarations to precede all code

Reply:
checkpatch seems to be allowing that, actually.  Not sure if that's a bug or an intended feature, but local declarations IMHO help a lot with clarity so I've been sneaking these by.

Reply:
Is there any way this can blow up the stack? This is supposed to be used in the scheduler, and it might be able to blow the kernel stack by creating a lot of threads

Reply:
max_depth is bounded to 2*log2(N) (it's a balanced tree), so not really.   Creating a billion threads would require 240 bytes of stack to traverse.